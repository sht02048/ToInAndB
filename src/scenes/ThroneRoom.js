import Renderer from "../graphics/Renderer";

import Boss from "../entities/enemies/Boss";
import Bot from "../entities/enemies/Bot";

class ThroneRoom extends Renderer {
  #bossWidth = 228;
  #spawnInterval = 60;

  constructor() {
    super();

    this.boss = new Boss({
      x: this.canvasWidth / 2 - this.#bossWidth / 2,
      y: -300,
    });
    this.botList = [];

    this.shouldBeDisplayed = false;
    this.hasStarted = false;
  }

  update() {
    this.boss.update();
    this.botList.forEach((bot) => {
      if (this.boss.isDestroyed) {
        bot.isDestroyed = true;
        bot.makeExplosionSound = false;
      }

      bot.update();
    });

    this.spawnBot();

    this.frame += 1;
  }

  render() {
    this.boss.render();
    this.botList.forEach((bot) => bot.render());
  }

  spawnBot() {
    if (
      !this.boss.shouldSpawnBot ||
      this.frame % this.#spawnInterval !== 0 ||
      this.boss.isDestroyed
    ) {
      return;
    }

    const leftBotX = this.boss.x + this.#bossWidth / 2 - 60;
    const rightBotX = this.boss.x + this.#bossWidth / 2 + 10;
    const botY = this.boss.y;

    const leftBot = new Bot({
      x: leftBotX,
      y: botY,
      isLeft: true,
    });
    const rightBot = new Bot({
      x: rightBotX,
      y: botY,
      isLeft: false,
    });

    this.botList.push(leftBot, rightBot);
  }

  setTarget(player) {
    this.boss.setTargetList(player);
    this.botList.forEach((bot) => bot.setTargetList(player));
  }

  setSceneTargetList() {
    const sceneTargetList = [this.boss];

    this.botList.forEach((bot) => sceneTargetList.push(bot));

    return sceneTargetList;
  }

  checkSceneStatus() {
    this.isDone = true;

    if (!this.boss.isDestroyed) {
      this.isDone = false;
    }

    this.botList.forEach((bot) => {
      if (!bot.isDestroyed && !bot.isVanished) {
        this.isDone = false;
      }
    });

    return this.isDone;
  }
}

export default ThroneRoom;
