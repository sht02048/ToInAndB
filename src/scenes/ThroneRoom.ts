import Renderer from "../graphics/Renderer";

import Player from "../entities/Player";
import Boss from "../entities/enemies/Boss";
import Bot from "../entities/enemies/Bot";

class ThroneRoom extends Renderer {
  #bossWidth = 228;
  #spawnInterval = 60;

  private boss: Boss;
  private botList: Bot[];
  private shouldBeDisplayed: boolean;
  private hasStarted: boolean;
  private isDone: boolean;

  constructor() {
    super();

    this.boss = new Boss({
      x: this.canvasWidth / 2 - this.#bossWidth / 2,
      y: -300,
    });
    this.botList = [];

    this.shouldBeDisplayed = false;
    this.hasStarted = false;
  }

  update(): void {
    this.boss.update();
    this.botList.forEach((bot: Bot) => {
      if (this.boss.isDestroyed) {
        bot.markDestroyed();
      }

      bot.update();
    });

    this.spawnBot();

    this.frame += 1;
  }

  render(): void {
    this.boss.render();
    this.botList.forEach((bot) => bot.render());
  }

  spawnBot(): void {
    if (
      !this.boss.shouldSpawnBot ||
      this.frame % this.#spawnInterval !== 0 ||
      this.boss.isDestroyed
    ) {
      return;
    }

    const leftBotX = this.boss.x + this.#bossWidth / 2 - 60;
    const rightBotX = this.boss.x + this.#bossWidth / 2 + 10;
    const botY = this.boss.y;

    const leftBot = new Bot({
      x: leftBotX,
      y: botY,
      isLeft: true,
    });
    const rightBot = new Bot({
      x: rightBotX,
      y: botY,
      isLeft: false,
    });

    this.botList.push(leftBot, rightBot);
  }

  setTarget(player: Player): void {
    this.boss.setTargetList(player);
    this.botList.forEach((bot) => bot.setTargetList(player));
  }

  setSceneTargetList(): (Boss | Bot)[] {
    const sceneTargetList = [this.boss, ...this.botList];

    return sceneTargetList;
  }

  checkSceneStatus(): boolean {
    this.isDone = true;

    if (!this.boss.isDestroyed) {
      this.isDone = false;
    }

    this.botList.forEach((bot: Bot) => {
      if (!bot.isDestroyed && !bot.isVanished) {
        this.isDone = false;
      }
    });

    return this.isDone;
  }
}

export default ThroneRoom;
