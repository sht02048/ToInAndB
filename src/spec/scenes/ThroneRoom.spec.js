import { beforeEach } from "vitest";
import ThroneRoom from "../../scenes/ThroneRoom";

describe("ThroneRoom", () => {
  const mainCanvas = document.createElement("canvas");
  const introCanvas = document.createElement("canvas");
  mainCanvas.id = "main-canvas";
  introCanvas.id = "intro-canvas";

  document.body.appendChild(mainCanvas);
  document.body.appendChild(introCanvas);

  let throneRoom;

  beforeEach(() => {
    throneRoom = new ThroneRoom();
  });

  it("should have proper functions", () => {
    expect(throneRoom.update).to.be.a("function");
    expect(throneRoom.render).to.be.a("function");
    expect(throneRoom.spawnBot).to.be.a("function");
    expect(throneRoom.setTarget).to.be.a("function");
    expect(throneRoom.setSceneTargetList).to.be.a("function");
    expect(throneRoom.checkSceneStatus).to.be.a("function");
  });

  it("should render boss", () => {
    throneRoom.render();

    const ctx = mainCanvas.getContext("2d");
    const events = ctx.__getEvents();

    expect(events).toMatchSnapshot();
  });

  it("should spawn bot", () => {
    const originalBotNumber = throneRoom.botList.length;
    throneRoom.frame = 60;
    throneRoom.boss.shouldSpawnBot = true;
    throneRoom.update();

    expect(originalBotNumber).to.not.equal(throneRoom.botList.length);
  });
});
