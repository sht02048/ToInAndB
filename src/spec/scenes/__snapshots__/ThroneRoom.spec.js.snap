// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ThroneRoom > should render boss 1`] = `
[
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 36,
      "dy": -300,
      "img": <img
        src="/src/assets/sprites/enemies/boss.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
]
`;
